test/Microsoft.ML.Tests/Scenarios/Api/TestApi.cs

```diff
@@ -183,6 +183,91 @@ public void TrainAveragedPerceptronWithCache()
             Assert.Equal(1, globalCounter);
         }
 
+        [Fact]
+        public void MetadataSupportInDataViewConstruction()
+        {
+            var data = ReadBreastCancerExamples();
+            var autoSchema = SchemaDefinition.Create(typeof(BreastCancerExample));
+
+            var mlContext = new MLContext(0);
+
+            // Create Metadata.
+            var kindFloat = "Testing float as metadata.";
+            var valueFloat = 10;
+            var coltypeFloat = NumberType.Float;
+            var kindString = "Testing string as metadata.";
+            var valueString = "Strings have value.";
+            var kindStringArray = "Testing string array as metadata.";
+            var valueStringArray = "I really have no idea what these features entail.".Split(' ');
+            var kindFloatArray = "Testing float array as metadata.";
+            var valueFloatArray = new float[] { 1, 17, 7, 19, 25, 0 };
+            var kindVBuffer = "Testing VBuffer as metadata.";
+            var valueVBuffer = new VBuffer<float>(4, new float[] { 4, 6, 89, 5 });
+
+            var metaFloat = new Microsoft.ML.Runtime.Api.MetadataInfo<float>(kindFloat, valueFloat, coltypeFloat);
+            var metaString = new Microsoft.ML.Runtime.Api.MetadataInfo<string>(kindString, valueString);
+
+            // Add Metadata.
+            var labelColumn = autoSchema[0];
+            var labelColumnWithMetadata = new SchemaDefinition.Column(mlContext, labelColumn.MemberName, labelColumn.ColumnType,
+                metadataInfos: new Microsoft.ML.Runtime.Api.MetadataInfo[] { metaFloat, metaString });
+
+            var featureColumnWithMetadata = autoSchema[1];
+            featureColumnWithMetadata.AddMetadata(kindStringArray, valueStringArray);
+            featureColumnWithMetadata.AddMetadata(kindFloatArray, valueFloatArray);
+            featureColumnWithMetadata.AddMetadata(kindVBuffer, valueVBuffer);
+
+            var mySchema = new SchemaDefinition { labelColumnWithMetadata, featureColumnWithMetadata };
+            var idv = mlContext.CreateDataView(data, mySchema);
+
+            Assert.True(idv.Schema[0].Metadata.Schema[kindFloat].Type == coltypeFloat);
```

Comment:
>Schema[kindFloat] [](start = 47, length = 17)

I would first assert that this is not null (here and in the next lines).

Reply:
Sorry, that's wrong because if the metadata of that name does not exist, this would throw.

Instead, you should assert the same way you do in line 226,227: Assert that the name of Schema[0] is kindFloat, and the type of Schema[0] is coltypeFloat.


---
In reply to: [240392932](https://github.com/dotnet/machinelearning/pull/1836#discussion_r240392932) [](ancestors = 240392932)

Reply:
Also, assert that Schema.Count is 2.

---
In reply to: [240396374](https://github.com/dotnet/machinelearning/pull/1836#discussion_r240396374) [](ancestors = 240396374,240392932)